src/file.rs

```diff
@@ -0,0 +1,180 @@
+// Copyright 2015-2021 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+/// Utilites for db file.
+
+use std::sync::atomic::{AtomicU64, AtomicBool, Ordering};
+use parking_lot::{RwLockUpgradableReadGuard, RwLock};
+use crate::error::Result;
+use crate::table::TableId;
+
+#[cfg(target_os = "linux")]
+fn disable_read_ahead(file: &std::fs::File) -> Result<()> {
+    use std::os::unix::io::AsRawFd;
+    let err = unsafe { libc::posix_fadvise(file.as_raw_fd(), 0, 0, libc::POSIX_FADV_RANDOM) };
+    if err != 0 {
+        Err(std::io::Error::from_raw_os_error(err))?
+    } else {
+        Ok(())
+    }
+}
+
+#[cfg(target_os = "macos")]
+fn disable_read_ahead(file: &std::fs::File) -> Result<()> {
+	use std::os::unix::io::AsRawFd;
+	if unsafe { libc::fcntl(file.as_raw_fd(), libc::F_RDAHEAD, 0) } != 0 {
+		Err(std::io::Error::last_os_error())?
+	} else {
+		Ok(())
+	}
+}
+
+#[cfg(not(any(target_os = "macos", target_os = "linux")))]
+fn disable_read_ahead(_file: &std::fs::File) -> Result<()> {
+	Ok(())
+}
+
+// `File::sync_data` uses F_FULLSYNC fcntl on MacOS. It it supposed to be
+// the safest way to make sure data is fully persisted. However starting from
+// MacOS 11.0 it severely degrades parallel write performance, even when writing to
+// other files. Regular `fsync` is good enough for our use case.
+// SSDs used in modern macs seem to be able to flush data even on unexpected power loss.
+// We performed some testing with power shutdowns and kernel panics on both mac hardware
+// and VMs and in all cases `fsync` was enough to prevent data corruption.
+#[cfg(target_os = "macos")]
+fn fsync(file: &std::fs::File) -> Result<()> {
+	use std::os::unix::io::AsRawFd;
+	if unsafe { libc::fsync(file.as_raw_fd()) } != 0 {
+		Err(std::io::Error::last_os_error())?
+	} else {
+		Ok(())
+	}
+}
+
+#[cfg(not(target_os = "macos"))]
+fn fsync(file: &std::fs::File) -> Result<()> {
+	file.sync_data()?;
+	Ok(())
+}
+
+
+const GROW_SIZE: u64 = 256 * 1024;
+
+pub struct DbFile {
```

Comment:
Let's call it `TableFile`. `DbFile` sounds like a generic database file. This is specific for storing fixed size records though.